File: iOS/Spaces V1.5/Models/SpaceAnnouncement.swift
import Foundation

enum AnnouncementAttachmentKind: String, Codable, CaseIterable, Hashable {
    case image
    case video
    case file
    case link

    var title: String {
        switch self {
        case .image: "Image"
        case .video: "Video"
        case .file: "File"
        case .link: "Link"
        }
    }

    var icon: String {
        switch self {
        case .image: "photo.fill"
        case .video: "video.fill"
        case .file: "doc.fill"
        case .link: "link"
        }
    }
}

struct AnnouncementAttachment: Identifiable, Codable, Hashable {
    let id: String
    let kind: AnnouncementAttachmentKind
    let title: String
    let urlString: String?
    let storagePath: String?
    let nonce: String?
    let mimeType: String?
    let fileSize: Int?
    let uploadedBy: String?

    init(
        id: String = UUID().uuidString,
        kind: AnnouncementAttachmentKind,
        title: String,
        urlString: String? = nil,
        storagePath: String? = nil,
        nonce: String? = nil,
        mimeType: String? = nil,
        fileSize: Int? = nil,
        uploadedBy: String? = nil
    ) {
        self.id = id
        self.kind = kind
        self.title = title
        self.urlString = urlString
        self.storagePath = storagePath
        self.nonce = nonce
        self.mimeType = mimeType
        self.fileSize = fileSize
        self.uploadedBy = uploadedBy
    }
}

enum AnnouncementReferenceKind: String, Codable, CaseIterable, Hashable {
    case event
    case note
    case list
    case media
    case file

    var title: String {
        switch self {
        case .event: "Event"
        case .note: "Note"
        case .list: "List"
        case .media: "Media"
        case .file: "File"
        }
    }

    var icon: String {
        switch self {
        case .event: "calendar"
        case .note: "note.text"
        case .list: "checklist"
        case .media: "photo.on.rectangle"
        case .file: "folder.fill"
        }
    }

    var emoji: String {
        switch self {
        case .event: "📅"
        case .note: "📝"
        case .list: "✅"
        case .media: "📷"
        case .file: "📁"
        }
    }
}

struct AnnouncementReference: Identifiable, Codable, Hashable {
    let id: String
    let kind: AnnouncementReferenceKind
    let targetID: String
    let title: String
    let subtitle: String?

    init(
        id: String = UUID().uuidString,
        kind: AnnouncementReferenceKind,
        targetID: String,
        title: String,
        subtitle: String? = nil
    ) {
        self.id = id
        self.kind = kind
        self.targetID = targetID
        self.title = title
        self.subtitle = subtitle
    }
}

struct AnnouncementReaction: Identifiable, Codable, Hashable {
    var id: String { emoji }
    let emoji: String
    var userIDs: Set<String>

    var count: Int { userIDs.count }
}

struct AnnouncementComment: Identifiable, Codable, Hashable {
    let id: String
    let authorID: String
    let authorName: String
    let body: String
    let createdAt: Date

    init(
        id: String = UUID().uuidString,
        authorID: String,
        authorName: String,
        body: String,
        createdAt: Date = Date()
    ) {
        self.id = id
        self.authorID = authorID
        self.authorName = authorName
        self.body = body
        self.createdAt = createdAt
    }
}

struct SpaceAnnouncement: Identifiable, Codable, Hashable {
    let id: String
    let spaceID: String
    var title: String
    var body: String
    let authorID: String
    let authorName: String
    let createdAt: Date
    var updatedAt: Date
    var isPinned: Bool
    var expiresAt: Date?
    var commentsEnabled: Bool
    var attachments: [AnnouncementAttachment]
    var references: [AnnouncementReference]
    var reactions: [AnnouncementReaction]
    var comments: [AnnouncementComment]

    var isExpired: Bool {
        guard let expiresAt else { return false }
        return expiresAt <= Date()
    }
}

File: Android/app/src/main/java/com/arcinteractive/spaces/data/model/SpaceAnnouncement.kt
package com.arcinteractive.spaces.data.model

import java.util.Date
import java.util.UUID

enum class AnnouncementAttachmentKind(val title: String, val emoji: String) {
    Image("Image", "\uD83D\uDDBC\uFE0F"),
    Video("Video", "\uD83C\uDFA5"),
    File("File", "\uD83D\uDCC4"),
    Link("Link", "\uD83D\uDD17")
}

data class AnnouncementAttachment(
    val id: String = UUID.randomUUID().toString(),
    val kind: AnnouncementAttachmentKind,
    val title: String,
    val url: String? = null,
    val storagePath: String? = null,
    val nonce: String? = null,
    val mimeType: String? = null,
    val fileSize: Long? = null,
    val uploadedBy: String? = null
)

enum class AnnouncementReferenceKind(val title: String, val emoji: String) {
    Event("Event", "\uD83D\uDCC5"),
    Note("Note", "\uD83D\uDCDD"),
    ListItem("List", "✅"),
    Media("Media", "\uD83D\uDCF7"),
    File("File", "\uD83D\uDCC1")
}

data class AnnouncementReference(
    val id: String = UUID.randomUUID().toString(),
    val kind: AnnouncementReferenceKind,
    val targetId: String,
    val title: String,
    val subtitle: String? = null
)

data class AnnouncementReaction(
    val emoji: String,
    val userIds: Set<String>
)

data class AnnouncementComment(
    val id: String = UUID.randomUUID().toString(),
    val authorId: String,
    val authorName: String,
    val body: String,
    val createdAt: Date = Date()
)

data class SpaceAnnouncement(
    val id: String = UUID.randomUUID().toString(),
    val spaceId: String,
    val title: String,
    val body: String,
    val authorId: String,
    val authorName: String,
    val createdAt: Date = Date(),
    val updatedAt: Date = Date(),
    val isPinned: Boolean = false,
    val expiresAt: Date? = null,
    val commentsEnabled: Boolean = true,
    val attachments: List<AnnouncementAttachment> = emptyList(),
    val references: List<AnnouncementReference> = emptyList(),
    val reactions: List<AnnouncementReaction> = emptyList(),
    val comments: List<AnnouncementComment> = emptyList()
) {
    val isExpired: Boolean
        get() = expiresAt?.let { !it.after(Date()) } ?: false
}
